src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java

```diff
@@ -0,0 +1,210 @@
+            } catch (final MalformedURLException e) {
+                // ignore
+            }
+        }
+        return open(new File(uri));
+    }
+
+    /**
+     * creates a VCF iterator from a File
+     * 
+     * @param file
+     *            the file (can be bcf, vcf, vcf.gz)
+     * @return the VCFIterator
+     * @throws IOException
+     */
+    public VCFIterator open(final File file) throws IOException {
+        IOUtil.assertFileIsReadable(file);
+        return open(new FileInputStream(file));
+    }
+
+    /** implementation of VCFIterator, reading VCF */
+    private static class VCFReaderIterator extends AbstractIterator<VariantContext>implements VCFIterator {
+        /** delegate input stream */
+        private final InputStream inputStream;
+        /** VCF codec */
+        private final VCFCodec codec = new VCFCodec();
+        /** VCF header */
+        private final VCFHeader vcfHeader;
+        /** Iterator over the lines of the VCF */
+        private final AsciiLineReaderIterator lineIterator;
+
+        VCFReaderIterator(final InputStream inputStream) {
+            this.inputStream = inputStream;
+            this.lineIterator = new AsciiLineReaderIterator(new AsciiLineReader(inputStream));
+            this.vcfHeader = (VCFHeader) this.codec.readActualHeader(this.lineIterator);
+        }
+
+        @Override
+        public VCFHeader getFileHeader() {
+            return this.vcfHeader;
+        }
+
+        @Override
+        protected VariantContext advance() {
+            return this.lineIterator.hasNext() ?
+                    this.codec.decode(this.lineIterator.next()) :
+                    null;
+        }
+
+        @Override
+        public void close() {
+            CloserUtil.close(this.lineIterator);
+            CloserUtil.close(this.inputStream);
+        }
+    }
+
+    /** implementation of VCFIterator, reading BCF */
+    private static class BCFInputStreamIterator extends AbstractIterator<VariantContext>implements VCFIterator {
+        /** the underlying input stream */
+        private final PositionalBufferedStream inputStream;
+        /** BCF codec */
+        private final BCF2Codec codec = new BCF2Codec();
+        /** the VCF header */
+        private final VCFHeader vcfHeader;
+
+        BCFInputStreamIterator(final InputStream inputStream) {
+            this.inputStream = new PositionalBufferedStream(inputStream);
+            this.vcfHeader = (VCFHeader) this.codec.readHeader(this.inputStream).getHeaderValue();
+        }
+
+        @Override
+        public VCFHeader getFileHeader() {
+            return this.vcfHeader;
+        }
+
+        @Override
+        protected VariantContext advance() {
+            return this.codec.isDone(this.inputStream) ?
+                    null :
+                    this.codec.decode(this.inputStream);
+        }
+
+        @Override
+        public void close() {
+            this.inputStream.close();
+        }
+    }
+
```

Comment:
@yfarjoun back to you:

in Ioutils, I've added:
```

public static boolean isGZIPInputStream(BufferedInputStream in);
public static InputStream mayBeGZippedInputStream(InputStream in);
```

and I've added some tests for those two functions in IoUtilTest.